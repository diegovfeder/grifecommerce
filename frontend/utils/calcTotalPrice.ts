import { CartItemProps } from '../@types/commonTypes';

export default function calcTotalPrice(cart: CartItemProps[]) {
	return cart.reduce((tally, cartItem) => {
		if (!cartItem.product) return tally;
		return tally + cartItem.quantity * cartItem.product.price;
	}, 0);
}
