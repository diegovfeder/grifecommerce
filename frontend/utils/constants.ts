export const TEXT_NO_DESCRIPTION = '- No description available';

export const NUM_TOTAL_PRODUCTS_PER_PAGE = 2;
