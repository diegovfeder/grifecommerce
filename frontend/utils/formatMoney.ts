const formatMoney = (amount = 0) => {
	const options = {
		style: 'currency',
		currency: 'BRL',
		minimumFractionDigits: 2,
	};

	if (amount % 100 === 0) {
		options.minimumFractionDigits = 2;
	}

	const formatter = Intl.NumberFormat('pt-BR', options);

	return formatter.format(amount / 100);
};

export default formatMoney;
