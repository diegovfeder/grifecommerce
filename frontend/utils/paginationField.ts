import { PRODUCTS_COUNT_QUERY } from '../gql/queries';

const paginationField = () => {
	return {
		keyArgs: false,
		// FIXME: Add proper types
		read(existing = [], { args, cache }: any) {
			const { take, skip } = args;
			const data = cache.readQuery({ query: PRODUCTS_COUNT_QUERY });
			const { productsCount } = data;
			const page = skip / take + 1;
			const pages = Math.ceil(productsCount / take);
			const items = existing.slice(skip, skip + take).filter(x => x);

			if (items.length && items.length !== take && page === pages) {
				return items;
			}
			if (items.length !== take) {
				return false;
			}
			if (items.length) {
				return items;
			}
			return false;
		},
		// FIXME: Add proper types
		merge(existing: any, incoming: any, { args }: any) {
			const { skip } = args;
			const merged = existing ? existing.slice(0) : [];
			for (let i = skip; i < skip + incoming.length; ++i) {
				merged[i] = incoming[i - skip];
			}
			return merged;
		},
	};
};

export default paginationField;
