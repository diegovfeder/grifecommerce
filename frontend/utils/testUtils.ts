// eslint-disable-next-line import/no-extraneous-dependencies
import casual from 'casual';
import { PRODUCTS_COUNT_QUERY } from '../gql/queries';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
	// __typename: 'Item',
	id: '718b7ac6-7cf1-47e7-b1de-c4a13ca92f2d',
	name: 'GRIFE Sample Pack - (Yellow)',
	description:
		'This our first sample pack for summer beats in a sunny day, when all is yellow :)',
	price: 5000,
	// user: null,
	photo: {
		// FIXME:
		// id: 'abc123',
		// altText: 'dogs are best',
		image: {
			publicUrlTransformed: '/samplepack_yellow.jpg',
		},
	},
});

const fakeUser = (rest?: any) => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
	...rest,
});

const fakeOrderItem = () => ({
	__typename: 'OrderItem',
	id: casual.uuid,
	image: {
		image: `${casual.word}.jpg`,
	},
	name: casual.words(),
	price: 4234,
	quantity: 1,
	description: casual.words(),
});

const fakeOrder = () => ({
	__typename: 'Order',
	id: 'ord123',
	charge: 'ch_123',
	total: 40000,
	items: [fakeOrderItem(), fakeOrderItem()],
	createdAt: '2022-12-11T20:16:13.797Z',
	user: fakeUser(),
});

const fakeCartItem = (rest?: any) => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 3,
	product: fakeItem(),
	user: fakeUser(),
	...rest,
});

// FIXME: Properly type this class
// Fake LocalStorage
class LocalStorageMock {
	store: { [key: string]: any } = {};
	constructor() {
		this.store = {};
	}

	clear() {
		this.store = {};
	}

	getItem(key: string) {
		return this.store[key] || null;
	}

	setItem(key: string, value: string) {
		this.store[key] = value.toString();
	}

	removeItem(key: string) {
		delete this.store[key];
	}
}

function makePaginationMocksFor(length: number) {
	return [
		{
			request: { query: PRODUCTS_COUNT_QUERY },
			result: {
				data: {
					_allProductsMeta: {
						count: length,
					},
					itemsConnection: {
						__typename: 'aggregate',
						aggregate: {
							count: length,
							__typename: 'count',
						},
					},
				},
			},
		},
	];
}

export {
	makePaginationMocksFor,
	LocalStorageMock,
	fakeItem,
	fakeUser,
	fakeCartItem,
	fakeOrder,
	fakeOrderItem,
};
