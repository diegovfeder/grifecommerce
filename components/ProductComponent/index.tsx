/* eslint-disable @next/next/link-passhref */
import React from 'react';
import { Container } from './styles';
import Image from 'next/image';
import Link from 'next/link';
import StyledProduct from '../styled/StyledProduct';
import StyledTitle from '../styled/StyledTitle';
import StyledPriceTag from '../styled/StyledPriceTag';
import formatMoney from 'util/formatMoney';
import DeleteProduct from 'components/DeleteProduct';

interface IProductComponent {
	product: {
		id: string;
		name: string;
		price: number;
		description: string;
		photo: {
			image: {
				publicUrlTransformed: string;
			};
		};
	};
}

const ProductComponent = ({ product }: IProductComponent) => {
	return (
		<Container>
			<StyledProduct>
				<Image
					src={product?.photo?.image?.publicUrlTransformed}
					alt={product.name}
					width={320}
					height={320}
				/>
				<StyledTitle>
					<Link href={`/product/${product.id}`}>{product.name}</Link>
				</StyledTitle>
				<StyledPriceTag>{formatMoney(product.price)}</StyledPriceTag>
				<p>{product.description}</p>
				<div className="buttonList">
					<Link
						href={{
							pathname: `/update/${product.id}`,
						}}
					>
						Edit üìù
					</Link>
					{/* TODO: Create Shopping Cart */}
					{/* <AddToCart id={product.id} /> */}
					{/* FIXME: When I delete a Product, a ProductImage stays in db. Shouldn't I...
					 clean that image if it isn't being used in another Product?  */}
					<DeleteProduct id={product.id}>Delete</DeleteProduct>
				</div>
			</StyledProduct>
		</Container>
	);
};

export default ProductComponent;
