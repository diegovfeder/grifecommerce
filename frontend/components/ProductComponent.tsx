import React from 'react';
import Image from 'next/image';
import Link from 'next/link';
import StyledItem from './styles/StyledItem';
import StyledTitle from './styles/StyledTitle';
import StyledPriceTag from './styles/StyledPriceTag';
import formatMoney from '../utils/formatMoney';
import DeleteProduct from './DeleteProduct';
import AddToCartButton from './AddToCartButton';
import { ProductProps } from '../types/commonTypes';
import { SupremeDescription } from './styles/Supreme';
import { LoadingLabel, LoadingSkeleton } from './loading';

const ProductComponent = ({
	product,
	loading,
}: {
	product: ProductProps;
	loading?: boolean;
}) => {
	// FIXME:
	// Where is this query coming from?
	// Why does it take so long for the image to load?..
	// Pass down loading...

	return (
		<StyledItem>
			{!!loading ? (
				<>
					<LoadingLabel />
					<LoadingSkeleton />
				</>
			) : (
				<Image
					src={product?.photo?.image?.publicUrlTransformed}
					alt={product?.name}
					width={'80%'}
					height={'80%'}
				/>
			)}
			<StyledTitle>
				<Link href={`/product/${product.id}`}>{product.name}</Link>
			</StyledTitle>
			<StyledPriceTag>{formatMoney(product.price)}</StyledPriceTag>
			<SupremeDescription>Description:</SupremeDescription>
			<p
				style={{
					marginTop: 0,
					marginBottom: 0,
					paddingTop: '1rem',
					paddingBottom: '2rem',
					fontSize: '1.5rem',
					maxHeight: '120px',
					overflow: 'scroll',
				}}
			>
				{product?.description || '- No description available'}
			</p>
			<div className="buttonList">
				{/* TODO: Move this Update Product Button  */}
				{/* TODO: Also change its UI and wrap in a custom component  */}
				<Link
					href={{
						pathname: '/update-product',
						query: {
							id: product.id,
						},
					}}
				>
					Update ‚úèÔ∏è
				</Link>
				<AddToCartButton id={product.id}>Add To Cart üõí</AddToCartButton>
				<DeleteProduct id={product.id}>Delete</DeleteProduct>
			</div>
		</StyledItem>
	);
};

export default ProductComponent;
