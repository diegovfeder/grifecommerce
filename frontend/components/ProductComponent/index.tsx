import StyledItem from '../styles/StyledItem';
import StyledTitle from '../styles/StyledTitle';
import { Container } from './styles';
import Image from 'next/image';
import Link from 'next/link';
import StyledPriceTag from '../styles/StyledPriceTag';
import formatMoney from 'util/formatMoney';

interface ProductComponentProps {
	product: {
		id: string;
		name: string;
		price: number;
		description: string;
		photo: {
			image: {
				publicUrlTransformed: string;
			};
		};
	};
}

function ProductComponent({ product }: ProductComponentProps) {
	return (
		<Container>
			<StyledItem>
				<Image
					src={product?.photo?.image?.publicUrlTransformed}
					alt={product.name}
					width={320}
					height={320}
				/>
				<StyledTitle>
					<Link href={`/product/${product.id}`}>{product.name}</Link>
				</StyledTitle>
				<StyledPriceTag>{formatMoney(product.price)}</StyledPriceTag>
				<p>{product.description}</p>
				{/* TODO: Add button for edit and delete item */}
			</StyledItem>
		</Container>
	);
}

export default ProductComponent;
