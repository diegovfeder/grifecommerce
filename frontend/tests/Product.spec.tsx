import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';

import { fakeItem } from '../utils/testUtils';
import mockedMutations from '../mocks/mutations';
import mockedQueries from '../mocks/queries';
import ProductComponent from '../components/ProductComponent';

const product = fakeItem();

describe('<ProductComponent/>', () => {
	it('renders component and matches the snapshot', () => {
		const { container, debug } = render(
			<MockedProvider
				mocks={[
					mockedQueries.currentUser,
					mockedMutations.addToCart,
					mockedMutations.deleteProduct,
				]}
				addTypename={false}
			>
				<ProductComponent product={product} />
			</MockedProvider>,
		);
		expect(container).toMatchSnapshot();
	});

	it('renders the price tag and title', () => {
		const { container } = render(
			<MockedProvider
				mocks={[
					mockedQueries.currentUser,
					mockedMutations.addToCart,
					mockedMutations.deleteProduct,
				]}
				addTypename={false}
			>
				<ProductComponent product={product} />
			</MockedProvider>,
		);
		const priceTag = screen.getByText('R$ 50,00');
		expect(priceTag).toBeInTheDocument();
		const link = container.querySelector('a');
		expect(link).toHaveAttribute(
			'href',
			'/product/718b7ac6-7cf1-47e7-b1de-c4a13ca92f2d',
		);
		expect(link).toHaveTextContent(product.name);
	});

	it('renders the product properly', () => {
		render(
			<MockedProvider
				mocks={[
					mockedQueries.currentUser,
					mockedMutations.addToCart,
					mockedMutations.deleteProduct,
				]}
				addTypename={false}
			>
				<ProductComponent product={product} />
			</MockedProvider>,
		);
		expect(screen.getByText(product.name)).toBeInTheDocument();
	});
});
