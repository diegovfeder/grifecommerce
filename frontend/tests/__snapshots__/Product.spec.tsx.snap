// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductComponent/> Renders and matches the snapshot 1`] = `
<div>
  <div
    class="StyledItem-sc-n8xnjs-0 dCHmEZ"
  >
    <div
      style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
    >
      <div
        style="box-sizing: border-box; display: block; max-width: 100%;"
      >
        <img
          alt=""
          aria-hidden="true"
          role="presentation"
          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMzIwIiBoZWlnaHQ9IjMyMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
          style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
        />
      </div>
      <noscript />
      <img
        alt="GRIFE Sample Pack - (Yellow)"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
      />
    </div>
    <h3
      class="StyledTitle-sc-xu1dk-0 AeOwD"
    >
      <a
        href="/product/718b7ac6-7cf1-47e7-b1de-c4a13ca92f2d"
      >
        GRIFE Sample Pack - (Yellow)
      </a>
    </h3>
    <span
      class="StyledPriceTag-sc-hnq0ox-0 hHCllp"
    >
      R$ 50,00
    </span>
    <p>
      This our first sample pack for summer beats in a sunny day, when all is yellow :)
    </p>
    <div
      class="buttonList"
    >
      <a
        href="/update-product?id=718b7ac6-7cf1-47e7-b1de-c4a13ca92f2d"
      >
        Update ✏️
      </a>
      <button
        class="AddToCartButton__StyledButtonLink-sc-er42vj-0 jJITkP"
        type="button"
      >
        Add To Cart 🛒
      </button>
      <button
        class="DeleteProduct__StyledButtonLink-sc-93lcvg-0 lfefWs"
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
